### ISSUE TITLE
elementwise scalar ops ignore strides on non-contiguous views, and `Ptree` uses unsupported syntax

### ISSUE THREAD

**Issue by nickbetteridge** [raven-ml/raven#202]
Hi

I was using opus to generate modules using Raven and hit a couple problems - this is the opus report:

# Raven `nx` — two issues found

Two independent problems:
(1) a **silent data-corruption bug** in elementwise scalar ops on
non-contiguous views, still present in the released alpha3; and (2) the current
`main` **fails to build** with the stock OCaml compiler.

## Environment

| | |
|---|---|
| OCaml | 5.4.1 |
| `nx` (installed / tested for bug 1) | `1.0.0~alpha3` (opam, `nx.cma` built 2026-06-24) |
| raven HEAD (tested for bug 2) | `eb60e100` |
| Platform | Linux x86-64 |

---

### Summary

`Nx.mul_s`, `Nx.div_s` (and `Nx.div` against a full tensor) operate on the raw
underlying buffer and **ignore the offset/strides of a non-contiguous view**. A
column slice such as `Nx.slice [Nx.A; Nx.R (3, 787)] m` is non-contiguous
(columns are interleaved in row-major storage), so a scalar op applied directly
to it reads the wrong memory — including bytes that belong to *other columns* —
and returns wrong values with no error.

Reductions (`Nx.max`/`Nx.min`/`Nx.sum`) **do** respect strides, which is what
makes this so dangerous: a sanity check via `max` reports the correct range while
the arithmetic underneath is silently corrupt.

### Why it is easy to miss

The corrupted op reads out-of-view buffer bytes, so **the bug is invisible
whenever those bytes are in the same value range as the view.** A synthetic
matrix filled with `i mod 256` normalizes "correctly" even at full size, because
every wrong read still lands in `[0, 255]`. The corruption only becomes
observable when out-of-view memory holds out-of-range values.

It is **not** size-dependent and **not** source-dependent: a fresh `Nx.create`
rebuild of the same values fails identically; a plain `reshape (arange …)` of the
same shape/strides passes only because its off-view bytes are in range.

### Minimal reproduction

```ocaml
let rmax v = Nx.max v |> Nx.item []

let () =
  let rows = 4000 and cols = 787 in
  (* pixels in [0,255] everywhere, EXCEPT column 0 holds a large sentinel *)
  let m =
    Nx.init Nx.float32 [| rows; cols |] (fun idx ->
        if idx.(1) = 0 then 24997.0
        else float_of_int ((idx.(0) + idx.(1)) mod 200))
  in
  let raw = Nx.slice [ Nx.A; Nx.R (3, 787) ] m in  (* excludes cols 0,1,2 *)

  Printf.printf "reduction max of slice        : %g\n" (rmax raw);
  (* => 199  (correct: strides respected, column 0 excluded) *)

  Printf.printf "mul_s (1/255) on raw slice max: %g\n"
    (rmax (Nx.mul_s raw (1.0 /. 255.0)));
  (* => 98.0275  == 24997/255  -> LEAKED column 0. WRONG. *)

  Printf.printf "mul_s (1/255) on contiguous   : %g\n"
    (rmax (Nx.mul_s (Nx.contiguous raw) (1.0 /. 255.0)))
  (* => 0.780392  == 199/255   (correct) *)
```

Observed output:

```
reduction max of slice        : 199
mul_s (1/255) on raw slice max: 98.0275
mul_s (1/255) on contiguous   : 0.780392
```

The `98.0275 = 24997/255` result proves the scalar op read column 0, which the
slice explicitly excluded.

### Real-world impact

Hit while normalizing MNIST-style pixel data loaded via `Nx_io.load_txt`: a
`[n; 787]` matrix (col 0 = id/large value, col 2 = label, cols 3..786 = pixels in
`[0,255]`). `mul_s (slice [A; R(3,787)] m) (1/255)` left pixels in `[0, 98]`
instead of `[0, 1]`. Downstream, a Bernoulli reconstruction cross-entropy went
negative (impossible) and a VAE diverged to neg-ELBO ≈ -3e10. The training code
was correct; the normalization was silently corrupt.

### Expected vs actual

- **Expected:** elementwise scalar ops honor the view's offset and strides (as
  reductions already do), or copy-on-read for non-contiguous inputs.
- **Actual:** they read the raw contiguous buffer, ignoring the view.

### Workaround

Wrap any non-contiguous view in `Nx.contiguous` before elementwise arithmetic.
Row-range slices (`Nx.slice [Nx.R (lo, hi); Nx.A]`) of a contiguous tensor are
themselves contiguous and are safe.

---

### Summary

At HEAD (`eb60e100`), building `nx` fails with a syntax error. The `Ptree`
module (added in `a7803f0f`, `packages/nx/lib/ptree.mli`) declares higher-rank
polymorphic **function arguments**:

```ocaml
val map : ('a 'b. ('a, 'b) Nx_effect.t -> ('a, 'b) Nx_effect.t) -> t -> t
```

Universally-quantified type variables in argument position are not accepted by
the OCaml compiler (higher-rank polymorphism is only allowed under record fields
or object methods). This is confirmed by a minimal standalone snippet:

```ocaml
module type S = sig val map : ('a 'b. int -> int) -> unit end
```

```
File "synt.mli", line 1, characters 34-35:
1 | module type S = sig val map : ('a 'b. int -> int) -> unit end
                                      ^
Error: Syntax error
```

### Build error

```
File "packages/nx/lib/ptree.mli", line 27, characters 16-17:
27 |   val map : ('a 'b. ('a, 'b) Nx_effect.t -> ('a, 'b) Nx_effect.t) -> t -> t
Error: Syntax error
File "packages/nx/lib/ptree.ml", line 9, characters 16-17:
Error: Syntax error
```

`packages/nx/lib/dune` has no `(preprocess (pps …))` stanza, so nothing rewrites
this syntax. To express higher-rank polymorphism, wrap the polymorphic function
in a record with a polymorphic field, e.g.:

```ocaml
type mapper = { f : 'a 'b. ('a, 'b) Nx_effect.t -> ('a, 'b) Nx_effect.t }
val map : mapper -> t -> t
```

(or an object method / first-class module), which the compiler accepts.

**Comment by tmattio** [comment]
Hi @nickbetteridge, thanks for the report! I'll have a look at the bug. For the compilation failure, we now depend on OCaml 5.5 to benefit from modular explicits, which Rune and Kaun have been rewritten on.

**Comment by tmattio** [comment]
Bug no 1 should has been fixed in the new C backend. Thanks again for the reports!
